modules/jupyternotebook/main.nf

```diff
@@ -0,0 +1,87 @@
+options        = initOptions(params.options)
+params.parametrize = true
+params.implicit_params = true
+params.meta_params = true
+
+process JUPYTERNOTEBOOK {
+    tag "$meta.id"
+    label 'process_low'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['id']) }
+
+    //NB: You likely want to override this with a container containing all required
+    //dependencies for you analysis. The container at least needs to contain the
+    //yaml and rmarkdown R packages.
+    //TODO: what container to use as default image?
+    conda (params.enable_conda ? "ipykernel jupytext nbconvert papermill" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/YOUR-TOOL-HERE"
+    } else {
+        container "quay.io/biocontainers/YOUR-TOOL-HERE"
+    }
+
+    input:
+    tuple val(meta), path(notebook)
+    val(parameters)
+    path(input_files)
+
+    output:
+    tuple val(meta), path("*.html"), emit: report
+    path("artifacts/*"), emit: artifacts
+    path "*.version.txt"          , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
+    // Dump parameters to yaml file.
+    // Using a yaml file over using the CLI params because
+    //  * no issue with escaping
+    //  * allows to pass nested maps instead of just single values
+    def params_cmd = ""
+    def render_cmd = ""
+    if (params.parametrize) {
+        nb_params = [:]
+        if (params.implicit_params) {
+            nb_params["cpus"] = task.cpus
+            nb_params["artifact_dir"] = "artifacts"
+            nb_params["input_dir"] = "."
+        }
+        if (params.meta_params) {
+            nb_params["meta"] = meta
+        }
+        nb_params += parameters
+        params_cmd = dump_params_yml(nb_params)
+        render_cmd = "papermill -f .params.yml"
+    } else {
+        render_cmd = "papermill"
+    }
+
+    """
+    # Create output directory
+    mkdir artifacts
+
+    # Set parallelism for BLAS/MKL etc. to avoid over-booking of resources
+    export MKL_NUM_THREADS="${task.cpus}"
+    export OPENBLAS_NUM_THREADS="${task.cpus}"
+    export OMP_NUM_THREADS="${task.cpus}"
+    export NUMBA_NUM_THREADS="${task.cpus}"
+
+    # dump parameters to yaml
+    ${params_cmd}
+
+    # Convert notebook to ipynb using jupytext, execute using papermill, convert using nbconvert
+    jupytext --to notebook --output - --set-kernel - ${notebook}  \\
+        | ${render_cmd} \\
+        | jupyter nbconvert --stdin --to html --output ${notebook.baseName}.html
+
+    # TODO how to output versions of multiple tools?
+    echo \$(jupytext --version) > ${software}.version.txt
```

Comment:
The nbconvert and papermill versions would also be relevant. How do we handle modules that use multiple pieces of software?

Reply:
I could have sworn any of the processes that piped also had a line for printing each tool version, but I can't find a single example. 
The bwa module for example only prints the bwa version and not the samtools version. The output declaration catches all version.txt files. 

Looking at the template, it seems it doesn't handle this:
The lines https://github.com/nf-core/tools/blob/b48bce26b7da46cb71e4a8f78b68d9ceac579ca6/nf_core/pipeline-template/workflows/pipeline.nf#L94-L95 seem to assume a single file coming out of the `first` operator rather than a list of files (although flattening should take care of that). I'm also not seeing how different versions of a tools are supported here. The `map` after `groupTuple` only takes the first entry.

Reply:
I will prefer a function similar to R::sessionInfo() to export all the imported modules and their versions.

Reply:
see other discussion thread and nf-core/tools#1247